### ISSUE TITLE
Frozen Language Model adapter [ROUND 1]

### ISSUE THREAD

**Issue by balthazarneveu** [balthazarneveu/molecule-retrieval-using-nlp#8]
0/ Fine tune the whole model but add a very tiny LR on this model (unfortunately, not sure we can specify a LR modulation for a specific part of gradients, may be possible using hooks) :stop_button:  definitely not the first idea to test

:snowflake: We should "freeze" the vast majority of the language model weights.
For the LLM to specify in chemistry, we can:
1/ Freeze all the weights and add an adapter (a few NN layers) on top on the LLM outputs. 
- [x] Freeze all weights = allows speed up, train without adapter :zap:  `exp_101 =  FBERT+adapter+base GCN` 19%
- [x] Add adapter `exp_106 FBERT+adapter+base GCN` 25%
- [ ] Prestore LLM embeddings to disk :racehorse: ? may be a sketchy idea  (probably no choice for SciBERT)

`exp_101 =  FBERT+adapter+base GCN` 19%
`exp_106 = FBERT+adapter+base GCN` 25%
`exp_113  = F-Sci-BERT+adapter+bigger GCN` 39.7 %

- [x] Does the improvement come from the bigger GCN or from SciBERT?  :heavy_check_mark:  improvement comes from both

![image](https://github.com/balthazarneveu/molecule-retrieval-using-nlp/assets/41070742/a1279f5a-dbeb-4a85-9246-a215be88b903)

Regex WandB:  `(F|Sci)BERT.*$`

2/ Fine tune only a few layers
3/ Use LORA (low rank adaptation) to fine tune only a few parameters :gift: 

:x: other point to be careful with: C-O-H2O  chemistry language may suffer from inputs embeddings. -> #5 Need to explore the caption dataset  but there are probably no such strings in the dataset.
:x: we could fine tune the language model on chemistry by fine-tuning on the chemistry corpus (mask training like BERT) -> :bulb:  Use SciBERT (any pretrained language model is allowed)   

Using SciBERT takes a lot of time to fine tune as a whole.
Frozen-SciBERT is pretty lightweight to train. Larger batch sizes could be used, SciBERT embeddings could be pre-computed
![image](https://github.com/balthazarneveu/molecule-retrieval-using-nlp/assets/41070742/5a10e454-7ced-449b-8a52-150dd95d39c9)
